src/fsharp/FSharp.Build/Microsoft.FSharp.Targets

```diff
@@ -411,49 +410,72 @@ this file.
-        <_NewPortableSdkPath>$(_NewRootSdkLocation)\$(_PortableRelativeSuffix)</_NewPortableSdkPath>
+        <_NewPortableSdkPath>$(NewFSharpSdkLocation)\$(_PortableRelativeSuffix)</_NewPortableSdkPath>
       </PropertyGroup>
 
       <ItemGroup>
         <!-- Update references to `.NETCore\*\FSharp.Core.dll`. -->
         <Reference Condition="'%(Reference.Identity)' == 'FSharp.Core' and
-                              '%(Reference.HintPath)' == '$(_OldCoreSdkPath)' and
-                              Exists('$(_NewCoreSdkPath)')">
+                              '%(Reference.HintPath)' == '$(_OldCoreSdkPath)'">
           <HintPath>$(_NewCoreSdkPath)</HintPath>
         </Reference>
 
         <!-- Update references to `.NETFramework\v4.0\*\FSharp.Core.dll`. -->
         <Reference Condition="'%(Reference.Identity)' == 'FSharp.Core' and
-                              '%(Reference.HintPath)' == '$(_OldFrameworkSdkPath)' and
-                              Exists('$(_NewFrameworkSdkPath)')">
+                              '%(Reference.HintPath)' == '$(_OldFrameworkSdkPath)'">
           <HintPath>$(_NewFrameworkSdkPath)</HintPath>
         </Reference>
 
         <!-- Update references to `.Portable\*\FSharp.Core.dll`. -->
         <Reference Condition="'%(Reference.Identity)' == 'FSharp.Core' and
-                              '%(Reference.HintPath)' == '$(_OldPortableSdkPath)' and
-                              Exists('$(_NewPortableSdkPath)')">
+                              '%(Reference.HintPath)' == '$(_OldPortableSdkPath)'">
           <HintPath>$(_NewPortableSdkPath)</HintPath>
         </Reference>
       </ItemGroup>
     </Target>
 
+
+    <Target Name="RedirectTPReferenceToNewRedistributableLocation" BeforeTargets="RedirectFSharpCoreReferenceToNewRedistributableLocation">
+      <PropertyGroup>
+        <_OldRefAssemTPLocation>$(MSBuildProgramFiles32)\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\4.3.0.0\Type Providers\FSharp.Data.TypeProviders.dll</_OldRefAssemTPLocation>
+        <_OldSdkTPLocation>Framework\v4.0\FSharp.Data.TypeProviders.dll</_OldSdkTPLocation>
+
+        <NewFSharpCompilerLocation Condition=" '$(NewFSharpCompilerLocation)' == '' ">$(MSBuildThisFileDirectory)</NewFSharpCompilerLocation>
+      </PropertyGroup>
+
+      <ItemGroup>
+        <!-- Update references to FSharp.Data.TypeProviders.dll -->
+        <Reference Condition="'%(Reference.Identity)' == 'FSharp.Data.TypeProviders' and
+                              '%(Reference.HintPath)' == '$(_OldRefAssemTPLocation)'">
+          <HintPath>$(NewFSharpCompilerLocation)\FSharp.Data.TypeProviders.dll</HintPath>
+        </Reference>
+
+        <!-- Update references to FSharp.Data.TypeProviders.dll -->
+        <Reference Condition="'%(Reference.Identity)' == 'FSharp.Data.TypeProviders' and
+                               $([System.String]::new('%(Reference.HintPath)').Endswith('$(_OldSdkTPLocation)', System.StringComparison.OrdinalIgnoreCase)) ">
```

Comment:
Since we're rewriting references I'd like to be as explicit as possible.  Since there's a version number that we can't easily deal with, can this check also include `HintPath.StartsWith('$(MSBuildProgramFiles32)\whatever')`.

Reply:
Also add an `Exists()` clause here.